File: Backend-js/src/routes/admin_users.ts
import { Router } from 'express';
import { Types } from 'mongoose';
import { AdminUser } from '../models/index.js';
import { HttpError } from '../middleware/error.js';
import { ah } from '../util/asyncHandler.js';
import { logger } from '../logger.js';

const ROLE_TO_DB: Record<string, string> = {
  transportation_admin: 'transport-admin',
  super_admin: 'transport-super',
};
const ROLE_FROM_DB: Record<string, string> = {
  'transport-admin': 'transportation_admin',
  'transport-super': 'super_admin',
};

function serialise(doc: Record<string, any>) {
  const dbRole = doc.role ?? 'transport-admin';
  return {
    id: String(doc._id),
    name: doc.full_name ?? '',
    email: doc.email ?? '',
    phone: doc.phone ?? '',
    role: ROLE_FROM_DB[dbRole] ?? dbRole,
  };
}

export const adminUsersRouter = Router();

adminUsersRouter.get(
  '/',
  ah(async (_req, res) => {
    const docs = await AdminUser.find({}).lean();
    res.json(docs.map(serialise));
  })
);

adminUsersRouter.get(
  '/find-by-email',
  ah(async (req, res) => {
    const email = String(req.query.email ?? '').trim().toLowerCase();
    if (!email) throw new HttpError(422, 'email required');
    const doc = await AdminUser.findOne({ email }).lean();
    if (!doc) throw new HttpError(404, 'Admin user not found');
    res.json(serialise(doc));
  })
);

adminUsersRouter.post(
  '/',
  ah(async (req, res) => {
    const body = req.body ?? {};
    const role: string = body.role ?? 'transportation_admin';
    if (!(role in ROLE_TO_DB)) {
      throw new HttpError(422, `role must be one of ${JSON.stringify(Object.keys(ROLE_TO_DB).sort())}`);
    }
    const email = String(body.email ?? '').trim().toLowerCase();
    if (!email || !body.name) throw new HttpError(422, 'name and email are required');

    const existing = await AdminUser.findOne({ email }).lean();
    if (existing) {
      res.status(409).json({ detail: `${email} is already registered.`, existing_id: String(existing._id) });
      return;
    }
    const phone = String(body.phone ?? '').trim();
    if (phone) {
      const existingPhone = await AdminUser.findOne({ phone }).lean();
      if (existingPhone) {
        res.status(409).json({
          detail: `Phone number ${phone} is already registered.`,
          existing_id: String(existingPhone._id),
        });
        return;
      }
    }
    const doc = {
      full_name: String(body.name).trim(),
      email,
      phone,
      role: ROLE_TO_DB[role],
      google_id: email,
      created_at: new Date(),
    };
    const created = await AdminUser.create(doc);
    logger.info(`Admin user created: ${created._id} (${email})`);
    res.status(201).json(serialise(created.toObject()));
  })
);

adminUsersRouter.patch(
  '/:userId',
  ah(async (req, res) => {
    const { userId } = req.params;
    if (!Types.ObjectId.isValid(userId)) throw new HttpError(400, 'Invalid user id');
    const oid = new Types.ObjectId(userId);
    const body = req.body ?? {};
    const update: Record<string, unknown> = {};

    if (body.name !== undefined) {
      const name = String(body.name).trim();
      if (!name) throw new HttpError(422, 'name cannot be empty');
      if (name.length > 160) throw new HttpError(422, 'name too long');
      update.full_name = name;
    }

    if (body.phone !== undefined) {
      const phone = String(body.phone).trim();
      if (phone.length > 32) throw new HttpError(422, 'phone too long');
      update.phone = phone;
    }

    if (Object.keys(update).length === 0) {
      throw new HttpError(422, 'No fields to update');
    }

    const updated = await AdminUser.findOneAndUpdate({ _id: oid }, { $set: update }, { new: true }).lean();
    if (!updated) throw new HttpError(404, 'User not found');
    logger.info(`AdminUser updated: ${oid}`);
    res.json(serialise(updated));
  })
);

adminUsersRouter.delete(
  '/:userId',
  ah(async (req, res) => {
    const { userId } = req.params;
    let query: any;
    if (Types.ObjectId.isValid(userId)) {
      query = { $or: [{ _id: new Types.ObjectId(userId) }, { _id: userId }] };
    } else {
      query = { _id: userId };
    }
    const result = await AdminUser.deleteOne(query);
    if (result.deletedCount === 0) throw new HttpError(404, 'User not found');
    res.status(204).end();
  })
);

File: archive/Backend/app/routes/admin_users.py
import logging
from datetime import datetime, timezone

from bson import ObjectId
from bson.errors import InvalidId
from fastapi import APIRouter, HTTPException, Query
from fastapi.responses import JSONResponse
from pydantic import BaseModel

from app.config import settings
from app.services.mongodb_service import get_database

router = APIRouter(tags=["Admin Users"])
logger = logging.getLogger(__name__)

# Frontend uses "transportation_admin"; MongoDB schema requires "transport-admin".
ROLE_TO_DB   = {"transportation_admin": "transport-admin", "super_admin": "transport-super"}
ROLE_FROM_DB = {"transport-admin": "transportation_admin", "transport-super": "super_admin"}


def _serialise(doc: dict) -> dict:
    db_role = doc.get("role", "transport-admin")
    return {
        "id":    str(doc["_id"]),
        "name":  doc.get("full_name", ""),
        "email": doc.get("email", ""),
        "phone": doc.get("phone", ""),
        "role":  ROLE_FROM_DB.get(db_role, db_role),
    }


def _oid(id_str: str) -> ObjectId:
    try:
        return ObjectId(id_str)
    except InvalidId:
        raise HTTPException(status_code=400, detail="Invalid user id")


class AdminUserCreate(BaseModel):
    name: str
    email: str
    phone: str = ""
    role: str = "transportation_admin"


@router.get("/")
async def list_admin_users():
    db = get_database()
    docs = await db[settings.admin_users_collection].find({}).to_list(None)
    return [_serialise(d) for d in docs]


@router.get("/find-by-email")
async def find_admin_user_by_email(email: str = Query(...)):
    db = get_database()
    doc = await db[settings.admin_users_collection].find_one({"email": email.strip().lower()})
    if not doc:
        raise HTTPException(status_code=404, detail="Admin user not found")
    return _serialise(doc)


@router.post("/", status_code=201)
async def create_admin_user(body: AdminUserCreate):
    if body.role not in ROLE_TO_DB:
        raise HTTPException(status_code=422, detail=f"role must be one of {sorted(ROLE_TO_DB)}")

    email = body.email.strip().lower()
    db = get_database()

    existing = await db[settings.admin_users_collection].find_one({"email": email})
    if existing:
        return JSONResponse(status_code=409, content={"detail": f"{email} is already registered.", "existing_id": str(existing["_id"])})

    phone = body.phone.strip()
    if phone:
        existing_phone = await db[settings.admin_users_collection].find_one({"phone": phone})
        if existing_phone:
            return JSONResponse(status_code=409, content={"detail": f"Phone number {phone} is already registered.", "existing_id": str(existing_phone["_id"])})

    doc = {
        "full_name":  body.name.strip(),
        "email":      email,
        "phone":      phone,
        "role":       ROLE_TO_DB[body.role],
        "google_id":  email,
        "created_at": datetime.now(timezone.utc),
    }
    result = await db[settings.admin_users_collection].insert_one(doc)
    doc["_id"] = result.inserted_id
    logger.info(f"Admin user created: {result.inserted_id} ({email})")
    return _serialise(doc)


@router.delete("/{user_id}", status_code=204)
async def delete_admin_user(user_id: str):
    db = get_database()
    try:
        oid = ObjectId(user_id)
        query = {"$or": [{"_id": oid}, {"_id": user_id}]}
    except InvalidId:
        query = {"_id": user_id}
    result = await db[settings.admin_users_collection].delete_one(query)
    if result.deleted_count == 0:
        raise HTTPException(status_code=404, detail="User not found")
